internal/scan/selector.go

```diff
@@ -0,0 +1,151 @@
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// GenericSelector can be used to implement the "Select" part of the Selector
+// interface. The "*Selector" fields are treated as Kubernetes selectors, all
+// other fields are regular expressions for matching metadata.
+type GenericSelector struct {
+	Group              string `json:"group,omitempty"`
+	Version            string `json:"version,omitempty"`
+	Kind               string `json:"kind,omitempty"`
+	Namespace          string `json:"namespace,omitempty"`
+	Name               string `json:"name,omitempty"`
+	LabelSelector      string `json:"labelSelector,omitempty"`
+	AnnotationSelector string `json:"annotationSelector,omitempty"`
+}
+
+// Select reduces the supplied resource node slice by only returning those
+// nodes which match this selector.
+func (g *GenericSelector) Select(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	m, err := newMetaMatcher(g)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]*yaml.RNode, 0, len(nodes))
+	for _, n := range nodes {
+		if meta, err := n.GetMeta(); err != nil {
+			return nil, err
+		} else if !m.matchesMeta(meta) {
+			continue
+		}
+
+		if matched, err := n.MatchesLabelSelector(g.LabelSelector); err != nil {
+			return nil, err
+		} else if !matched {
+			continue
+		}
+
+		if matched, err := n.MatchesAnnotationSelector(g.AnnotationSelector); err != nil {
+			return nil, err
+		} else if !matched {
+			continue
+		}
+
+		result = append(result, n)
+	}
+
+	return result, nil
+}
+
+// metaMatcher matches metadata. This is an alternative to `types.SelectorRegex` from Kustomize.
+type metaMatcher struct {
+	namespaceRegex *regexp.Regexp
+	nameRegex      *regexp.Regexp
+	groupRegex     *regexp.Regexp
+	versionRegex   *regexp.Regexp
+	kindRegex      *regexp.Regexp
+}
+
+// newMetaMatcher creates a new matcher by compiling all of the regexps.
+func newMetaMatcher(g *GenericSelector) (m *metaMatcher, err error) {
+	m = &metaMatcher{}
+
+	m.namespaceRegex, err = compileAnchored(g.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	m.nameRegex, err = compileAnchored(g.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	m.groupRegex, err = compileAnchored(g.Group)
+	if err != nil {
+		return nil, err
+	}
+
+	m.versionRegex, err = compileAnchored(g.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	m.kindRegex, err = compileAnchored(g.Kind)
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// matchesMeta returns true if the supplied metadata is accepted by this matcher.
+func (m *metaMatcher) matchesMeta(meta yaml.ResourceMeta) bool {
+	if m.namespaceRegex != nil && !m.namespaceRegex.MatchString(meta.Namespace) {
+		return false
+	}
+	if m.nameRegex != nil && !m.nameRegex.MatchString(meta.Name) {
+		return false
+	}
+
+	if m.groupRegex != nil || m.versionRegex != nil {
+		group, version := "", meta.APIVersion
+		if pos := strings.Index(version, "/"); pos >= 0 {
+			group, version = version[0:pos], version[pos+1:]
+		}
```

Comment:
should we use https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime/schema#ParseGroupVersion instead?
ultimately the same, so indifferent on this suggestion